### PR TITLE
[Content Addressable] Support Content Addressable Gems in Remote Install Process

### PR DESCRIPTION
**Pull request by OughtPuts**
## TL;DR

  Adds support for resolving, downloading, and installing content-addressed (“skinny”) gems from a remote source.

  ## Description
  
  Addresses: https://github.com/ruby/rubygems/issues/9731

[Note - at the moment we also have the commit from the local install work on this branch as well as the first commit. That will be removed and this PR will be rebased once the local install work is merged into the feature branch.]

  Compact-index entries for skinny gems use the content address as the version suffix and provide the real platform through the `platform:` requirement.

  This change:

  - Separates the content address from the gem’s real platform.
  - Carries the content address through the resolver into the full specification.
  - Prefers a compatible skinny gem over fat and source variants.
  - Falls back to fat or source gems when the skinny gem is incompatible.
  - Downloads gems and development gemspecs using their content-addressed filenames.
  - Preserves distinct resolver candidates when their content addresses differ.

  This PR builds on the local content-addressed installation work.

  ## Testing

```
  ruby --disable-gems -Ilib:test \
    test/rubygems/test_gem_resolver_api_set.rb \
    test/rubygems/test_gem_resolver_api_specification.rb \
    test/rubygems/test_gem_resolver.rb \
    test/rubygems/test_gem_remote_fetcher.rb
```

  The tests cover:

  - Parsing content-addressed compact-index entries.
  - Selecting compatible skinny gems.
  - Falling back to fat or source gems.
  - Respecting Ruby and RubyGems requirements.
  - Downloading SHA-named gem archives.
  - Fetching SHA-named gemspecs for development dependencies.

  ## Top hatting

  1. Configure a gem server containing compatible skinny, fat, and source variants.
  2. Install the gem using this RubyGems checkout:

```
     ruby --disable-gems -Ilib exe/gem install GEM_NAME \
       --source SERVER_URL \
       --install-dir "$(mktemp -d)"
```

  3. Confirm the compatible SHA-named gem was downloaded and installed.
  4. Confirm the installed specification reports the real platform.
  5. Confirm the gem can be required.
  6. Repeat with no compatible skinny variant and confirm installation falls back to the fat or source gem.
  
**Tested the following locally, behaving as expected:** 
- Remote compact-index lookup.
- CA gem selection by Ruby ABI and platform.
- SHA-named gem download and installation.
- Correct installed SHA directory, gemspec, and cache names.
- Correct real platform and content address.
- Successful activation with require.
- Idempotent CA re-installation.
- Normal non-CA platform gem download, installation, and activation.

### CODE REVIEW

File: lib/rubygems/installer.rb [hunk 1]
```diff
@@ -966,6 +970,26 @@ def ensure_writable_dir(dir) # :nodoc:
...
+
+    filename = File.basename(path, ".gem")
+    token = filename.split("-").last
+    return nil unless Gem::BasicSpecification.content_address?(token)
+
+    require "digest"
+    digest = Digest::SHA256.file(path).hexdigest
+    raise Gem::InstallError, "content address mismatch for #{File.basename(path)}" unless digest.start_with?(token)
+    token
+  end
```
[the remark this answers is not in the document]
Comment (OughtPuts) [reply lib/rubygems/installer.rb]: I've had a look at some pre-release versions on RubyGems.org. I downloaded one, and it's filename is in this format: 

```
gitaly-eju-16.7.0.pre.1cf5071d5.gem
```

As we are splitting on the '-' characters, our token value would resolve to `16.7.0.pre.1cf5071d5.gem` which would fail the `content_address?` regex check.

[the remark this answers is not in the document]
Comment (OughtPuts) [reply lib/rubygems/installer.rb]: I've added a bit of hardening to the `derive_content_address_` method anyway though (and test cases) just to make sure we're trying to anticipate against the edge cases: 
- Normal filename match: if the actual filename equals spec.file_name, don’t treat a hex-looking suffix as a content address.
- Spec mismatch: require the actual filename to begin with the spec’s exact name-version- prefix.
- Known address mismatch: when the spec already has a content address, require the filename to match it.

**Review by jenshenny — APPROVED** [review]
